Realm.Shared/linq/RealmResultsVisitor.cs

```diff
@@ -179,44 +216,60 @@ internal override Expression VisitMethodCall(MethodCallExpression m)
+                    return MakeDefaultWithTransactionIfNeeded(_metadata);
                 }
-                if (m.Method.Name == "Single")  // same as unsorted First with extra checks
+                if (m.Method.Name.StartsWith("Single"))  // same as unsorted First with extra checks
                 {
                     RecurseToWhereOrRunLambda(m);  
                     var firstRowPtr = _coreQueryHandle.FindDirect(IntPtr.Zero);
-                    if (firstRowPtr == IntPtr.Zero)
-                        throw new InvalidOperationException("Sequence contains no matching element");
+                    if (firstRowPtr == IntPtr.Zero) {
+                        if (m.Method.Name == "Single")
+                            throw new InvalidOperationException("Sequence contains no matching element");
+                        Debug.Assert (m.Method.Name == "SingleorDefault");
+                        return MakeDefaultWithTransactionIfNeeded(_metadata);
+                    }
                     var firstRow = Realm.CreateRowHandle(firstRowPtr, _realm.SharedRealmHandle);
                     IntPtr nextIndex = (IntPtr)(firstRow.RowIndex+1);
                     var nextRowPtr = _coreQueryHandle.FindDirect(nextIndex);
                     if (nextRowPtr != IntPtr.Zero)
                         throw new InvalidOperationException("Sequence contains more than one matching element");
                     return Expression.Constant(_realm.MakeObjectForRow(_metadata, firstRow));
                 }
-                if (m.Method.Name == nameof(Queryable.ElementAt))
+                if (m.Method.Name.StartsWith("Last"))
                 {
-                    Visit(m.Arguments.First());
-                    var index = (int)ExtractConstantValue(m.Arguments.Last());
+                    RecurseToWhereOrRunLambda(m); 
 
-                    RowHandle row = null;
-                    if (OptionalSortDescriptorBuilder == null)
+                    var lastRowPtr = IntPtr.Zero;
+                    using (ResultsHandle rh = _realm.MakeResultsForQuery(_coreQueryHandle, OptionalSortDescriptorBuilder)) 
                     {
-                        var rowPtr = _coreQueryHandle.FindDirect((IntPtr)index);
-                        row = Realm.CreateRowHandle(rowPtr, _realm.SharedRealmHandle);
-                    }
-                    else
-                    {
-                        using (ResultsHandle rh = _realm.MakeResultsForQuery(_coreQueryHandle, OptionalSortDescriptorBuilder))
-                        {
-                            var rowPtr = rh.GetRow(index);
-                            row = Realm.CreateRowHandle(rowPtr, _realm.SharedRealmHandle);
-                        }
+                        var lastIndex = rh.Count() - 1;
+                        if (lastIndex >= 0)
+                            lastRowPtr = rh.GetRow(lastIndex);
                     }
+                    if (lastRowPtr != IntPtr.Zero)
```

Comment:
With the above methods, we first evaluate the `rowPtr == IntPtr.Zero` case and then the `!= IntPtr.Zero` case, so it's a little inconsistent here.